### PR TITLE
feat(payments): update PlanUpgradeDetails component content

### PR DESCRIPTION
**Pull request by sardesam**
## Because

- We need to show the customer what is owed and when (i.e. proration) on checkout.

## This pull request

- Updates `PlanUpgradeDetails` component and invoice preview to include proration amount.
- Adds a helper function to format for only plan intervals.
- Shows updated content when `useStripeImmediatelyInvoice` flag is set to `true`.
- Updates tests and stories when aforementioned flag is set to `true`, where applicable.

## Issue that this pull request solves

Closes FXA-6878

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (flodolo) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:16]:
```suggestion
## Frequency of payment for subscription plan change
```

Use a group-level comment (two dashes) if it applies to multiple strings, then reset it after the group with `##`. We should also make it clearer that it related to frequency of payment.

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (flodolo) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:13]: Why plural "charges"? If it's one-time, wouldn't that make it a single charge?

Also, should it be [one-time](https://grammarist.com/spelling/one-time-onetime/) instead of _one time_?

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (flodolo) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:14]: What's "Upgrade" here? Is it a noun, as in "Cost for the upgrade", or a verb?

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:13]: "charges" was in the mock-up, but you're right, it looks like a typo. And good catch. Should be "One-time". My bad. Thank you. Both have been corrected!

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:14]: I thought of it more as an adjective to describe the type of purchase. But there is a price to be added beside it, so then I agree it would be a noun. Do I add a comment above it to make that clear?

Comment (flodolo) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:14]: Yes, please. Something like `Upgrade is a noun, as in "cost for upgrade", and is followed by the price`.

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (flodolo) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:16]:
```suggestion
## Frequency of payment for subscription plan change

```

Leave an empty line after

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (flodolo) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:20]:
```suggestion
sub-update-interval-yearly = Yearly

##

```
Reset the group comment

Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:20]: TIL! Thank you @flodolo . I didn't know about resetting.

Reference: https://firefox-source-docs.mozilla.org/l10n/fluent/review.html#comments

**Review by flodolo — APPROVED** [review]

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx
Comment (StaberindeZA) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx:34]: At the moment `formatPlanInterval` returns different values depending on 1 and non-1 `interval_count`. However the l10n string only supports values for the case where `interval_count` is 1. This could lead to a mismatch between the l10n string and the fallback text.

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx:92]: I like the idea of giving extra context but I'm not quite following what you mean. Could you please elaborate?

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx]: Good catch! Fixed.

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx:92]: I am all for providing as much clarity as possible, but I'm not quite following how adding the product name would provide extra context? Could you please elaborate? As per flod's comment, I think the updated comment copy is meant to provide the necessary context.

Comment (StaberindeZA) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx:92]: I was just wondering if including the product name and the parentheses would give the translator better full context of how and where the translated string will be used. (At the moment the translator isn't aware of the product name or the parentheses in the string) As you said, l10n already reviewed this, so maybe it's not necessary or desired in this case.

Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx:92]: Oh! I see what you mean. That makes sense @flodolo what are your thoughts? I have to make some copy updates so I will need your approval again anyway when they are ready 🙏 .

Comment (pdehaan) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx:92]: Maybe I'm reading the code wrong, but above on L34 it looks like we're hardcoding `intervalCount: 1` to the `formatPlanInterval()`, which means we'd only get the four possible values `"daily"` | `"weekly"` | `"monthly"` | `"yearly"`.
Then we're trying to match the following localized ID:

```sh
`sub-update-interval-${formattedInterval}`
```

Why do we need to lowercase and then uppercase the first char if they're already localized w/ sentence casing in the .ftl files?

```ini
sub-update-interval-daily = Daily
sub-update-interval-weekly = Weekly
sub-update-interval-monthly = Monthly
sub-update-interval-yearly = Yearly
```

Ah&hellip; unless the string is the fallback value.

---

```sh
git grep "sub-update-interval-" | cat

packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx:                    `sub-update-interval-${formattedInterval}`,
packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:sub-update-interval-daily = Daily
packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:sub-update-interval-weekly = Weekly
packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:sub-update-interval-monthly = Monthly
packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:sub-update-interval-yearly = Yearly
```

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (pdehaan) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:20]: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx#L161-162 shows this using a `<span className="font-semibold">Upgrade</span>` instead of a `<strong>Upgrade</strong>`:
```html
<span className="font-semibold">Upgrade</span> (Prorated to end of period)
```

File: packages/fxa-payments-server/src/lib/formats.ts
Comment (sardesam) [reply packages/fxa-payments-server/src/lib/formats.ts:180]: Typo. Thank you! Fixed.

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx:34]: `interval_count` can be either 1 or non-1, which can possibly lead to error if passed in directly. Only 1 returns the adverb that we want, so in this particular case it was hardcoded.

File: packages/fxa-payments-server/src/lib/formats.ts
Comment (StaberindeZA) [inline packages/fxa-payments-server/src/lib/formats.ts:184]: I noticed this is missing some tests.

Also, since 1 is being hard coded, does this method need `intervalCount`? Removing it would simplify the function and tests a bit.

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx
Comment (bcolsson) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx]: (Bear with me as I'm not 100% sure I'm reading the code correctly)

Could you clarify what new strings will be appearing on the screen. From what I can see in the Jira ticket actual strings will look like the below in English:
`{Product Name} monthly`

We don't want to construct the string with code, the string for translation should really be what will appear on screen. This way translators can ensure the correct word order and formatting for their locale.

File: packages/fxa-payments-server/src/lib/formats.ts
Comment (sardesam) [reply packages/fxa-payments-server/src/lib/formats.ts:184]: Haven't done tests yet. I was waiting for the blocking ticket first before doing them altogether but good reminder, so thank you!

I think having a generic function for just the frequencies would be good. `formatPlanPricing` is more restrictive, as it includes pricing. But let me know your thoughts. I know a lot of things are changing for 3.0 so if it's not worth the extra effort INHO, then I can update as suggested.

**Comment by sardesam** [comment]
@flodolo re-requesting review. Copy had to be updated. This is the final copy. TYSM!

**Comment by bcolsson** [comment]
> @flodolo re-requesting review. Copy had to be updated. This is the final copy. TYSM!

Hi @sardesam , Flod is out on PTO (and he took over this ticket while I was on PTO) so I'll handle the review. 

I actually had a question and put it in this thread here:
https://github.com/mozilla/fxa/pull/15498#discussion_r1261521230

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx]: That is what Reino was saying. Agreed with you both. I'll update the string to include the product name and parentheses. 🙂

Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/PlanUpgradeDetails.tsx]: @bcolsson updated! Let me know if this works: https://github.com/mozilla/fxa/pull/15498/files#diff-626bf53978856be986cbc1a3048c8817e326e9a0bcb137ec4a7540a857e5f2df

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (bcolsson) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:22]:
```suggestion
## Checkout line item for subscription plan change listing the product name and frequency of payment 
## For example, a Mozilla VPN subscription charged monthly would appear as: Mozilla VPN (Monthly)
## Variables:
##   $name (String) - Name of the upgraded product (e.g. Mozilla VPN)

```
Individual comments (#) would only appear with the first string immediately after it, so moving the variable explanation to the group comment. 
Also added some additional context to the comment to make it easier for translators to picture the context.

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (bcolsson) [inline packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl:26]:
```suggestion
sub-update-new-plan-daily = { $name } (Daily)
sub-update-new-plan-weekly = { $name } (Weekly)
sub-update-new-plan-monthly = { $name } (Monthly)
sub-update-new-plan-yearly = { $name } (Yearly)
```

Fluent variables should have spaces separating it from the brackets.

Nit: It'd be nicer if the variable name were more descriptive (e.g. $productName instead of $name), again this would be for localizers to better picture how the string would appear.

**Review by bcolsson — CHANGES_REQUESTED** [review]
Thanks, this is looking good. I still have some comments, but having the whole string will allow all locales to translate this accurately!

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl]: Updated as per your suggestions! Thank you the additional info ✍️ . Good to know. Let me know!

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl]: Agreed. I was originally following earlier variable naming for consistency but I much prefer a more descriptive variable name myself. Updated to `$productName`. 😄 Let me know!

**Review by bcolsson — APPROVED** [review]
Approving, though there's one last minor fluent syntax issue it should be good to go after that fix.

File: packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl
Comment (sardesam) [reply packages/fxa-payments-server/src/routes/Product/SubscriptionUpgrade/en.ftl]: Fixed!

**Review by IvoJP — APPROVED** [review]

**Pull request merged**
